### ISSUE TITLE
tf.data.Dataset.shard_and_drop_remainder?

### ISSUE THREAD

**Issue by terrykong** [tensorflow/tensorflow#19952]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
      - yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow installed from (source or binary)**:
- **TensorFlow version (use command below)**:
       - v.1.8.0
- **Python version**: 
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem
I'm curious if there's anyway to accomplish a function like Dataset.shard_and_drop_remainder(num_shards, index) with the current machinery? If not, then I guess this is a feature request. The use case I have is, I'm training an RNN model with horovod and I've decided to bucket my variable length sequences, then batch them, and then shard to the number of workers, but I have no idea if there are enough batches at all my workers to perform an allreduce, which results in horovod complaining and failing in the allreduce. Ideally I would like the dataset to take "num_shards" and use that to check whether I have enough elements in the dataset to carry out another global step.

The solution I've come up with now is to iterate once to count the number of batches and then use Dataset.take((num_batches//num_workers)*num_workers) to ensure the number of batches is divisible by the number of workers.

**Comment by terrykong** [comment]
I've thought of using Dataset.batch (and even Dataset.padded_batch), to batch num_workers and then tf.contrib.data.unbatch() them, but the shapes don't match, i.e., [batch_size, max_seq_len_1] and [batch_size, max_seq_len_2], not to mention it is a lot of extra work to batch() -> unbatch -> rebatch().

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cy89** [comment]
@mrry could you please comment on whether there's a way to do what @terrykong wants, or if this is a feature request, whether we might do it, or we should mark it "Community Support"?

**Comment by mrry** [comment]
I think this should be possible with the new batching operations that @jsimsa is working on.

**Comment by jsimsa** [comment]
Hello @terrykong, the workaround you used is currently the best known solution. However, as alluded to by @mrry, I am working and a couple of new transformations that will allow you do to the following:

```
dataset = dataset.batch(batch_size).window(num_workers).filter(lambda x: tf.equal(count(x), num_workers)).flat_map(lambda x: x)
```

The new transformations are:
`window` -- which combined elements of the input dataset into windows of nested dataset
`count` -- which assumes the input is a dataset and counts the number of elements of its elements

The design is expect to undergo an external review soon and I will notify you when it does.

**Comment by terrykong** [comment]
Very cool @jsimsa! I look forward to your changes :)

**Comment by jsimsa** [comment]
@terrykong see the RFC [here](https://github.com/tensorflow/community/pull/5#discussion_r205701432).

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jsimsa: It has been 18 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jsimsa: It has been 34 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jsimsa: It has been 49 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jsimsa: It has been 64 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jsimsa: It has been 79 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jsimsa: It has been 94 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jsimsa: It has been 109 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
